### ISSUE TITLE
Support a "warning" level for linting

### ISSUE THREAD

**Issue by lengau** [charliermarsh/ruff#1256]
It would be really nice if, like `ignore`, we could select specific issues to turn into warnings rather than errors. That is, it would still output a warning to `stdout`, (perhaps separating errors and warnings), but it would still return with exit status 0. 

### Use case:

A large, older codebase that contains a lot of linting issues regarding specific rules. (For example, a codebase that predates type annotations.) By allowing warnings, we can make these issues visible to developers for fixing as they lint their changes, but we don't have to have a separate configuration for our pre-commit hooks, CI tools, etc. to ensure they ignore these issues.

**Comment by charliermarsh** [comment]
Yeah, this would be nice to have. It would also be helpful for editor integrations, since the LSP can differentiate between warnings and errors.

**Comment by charliermarsh** [comment]
Any thoughts on what the API should look like?

ESLint lets you map from rule name to level, like:

```
...
"eqeqeq": "error",
"curly": "warning",
...
```

Which makes sense, though they also use that system to turn errors on and off -- and we instead have the Flake8-like `ignore` and `select` system.

The most basic thing would be a map from check code prefix to level, like:

```toml
[tool.ruff.severity]
"F401" = "warning"
"F481" = "error"
# Probably also want to support prefixes.
"E" = "warning"
```

**Comment by lengau** [comment]
Not really sure... Naively, it appears to me that the most consistent way to configure it would be to have `warn` and `extend-warn` (or `warning` & `extend-warning`).

[This file](https://github.com/canonical/craft-parts/pull/337/files#diff-50c86b7ed8ac2cf95bd48334961bf0530cdc77b5a56f852c5c61b89d735fd711) contains what I'm doing currently for the project that inspired this request.

I enabled more linting than we previously had (everything I viewed as potentially appropriate for this app), but I put everything that raised an error into the ignore field. With this feature, lines 78-102 would've been in a `warn` section instead of inside `ignore`, with the goal of eventually emptying the `warn` section by consensus.

**Comment by edgarrmondragon** [comment]
fwiw `bandit` has four severity levels: undefined, low, medium and high.

https://github.com/PyCQA/bandit/blob/af9f8dc09f7239f3c3481177a24a72dcb41ed450/bandit/core/constants.py#L8

**Comment by jfmengels** [comment]
I would recommend against the implementation of warning (or severity levels). I wrote about why [in this blog post](https://jfmengels.net/disable-comments/) (and also in a as-of-yet-unreleased talk).

In my opinion, it's one of those features (along with disable comments) that gets copied over and over mostly because other linters have them. The fact that no linter docs ever recommend when to use these features is a sign of that in my opinion.

I can potentially see a benefit for this while seeing the errors in an editor, but it shouldn't IMO impact the behavior of the linter.

In [`elm-review`](https://elm-review.com) there are no severity levels (nor disable comments): rules are enforced or they are not. Though to be taken with a grain of salt: the linter also has a lot less false positives than is probably possible for Python because of the target language.

> A large, older codebase that contains a lot of linting issues regarding specific rules

This is exactly the kind of problem that the proposal in #1149 aims to solve, and is a much better solution.

**Comment by jacobmischka** [comment]
> I can potentially see a benefit for this while seeing the errors in an editor, but it shouldn't IMO impact the behavior of the linter.

How would the editor show the proper type of diagnostic to the user if the linter doesn't provide that information?

**Comment by jfmengels** [comment]
If by "type of diagnostic" you mean warning or error, it wouldn't be able to. I meant to say that this feature request could be useful for editors for that purpose. But that otherwise severity levels is not useful or recommended.

**Comment by charliermarsh** [comment]
@not-my-profile - I'm considering adding this, but since you've been working on the violation API a lot, I wanted to get your feedback first. Here's a proposal.

- Add a `Severity` enum, with `Warning` and `Error` for now (which map to the LSP's `DiagnosticSeverity`).
- Add a const on the `Violation` trait that allows rules to define their default severity, like `AUTOFIX`. All rules will be `Error` for now, so we could consider omitting this at the moment.
- Add a `warnings = [...]` setting in `pyproject.toml`, that's functionally similar to `fixable`. All rules are errors by default, so users can downgrade them to `warnings` like so: `warnings = ["B"]`.
- In the CLI output, and in `ruff-lsp`, differentiate errors from warnings.

This wouldn't include making it possible to run Ruff and _only_ show warnings, or _only_ show errors (e.g., `ruff --select warnings`), but we could do that in the future.

**Comment by not-my-profile** [comment]
Thanks for consulting with me :) I would hold off on adding any such constants to the `Violation` trait till we have implemented rule categories #1774 ... I am not sure if adding constants to `Violation` is the best way to model non-autofix-related rule metadata ... I'll have to think about it.

> Add a Severity enum, with Warning and Error for now (which map to the LSP's DiagnosticSeverity).

LSP also supports [other levels](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#diagnosticSeverity):

* error
* warning
* information
* hint

I think it might make more sense to support all of them via a configuration like:

```toml
[tool.ruff.severities]
B = "information"
```

> e.g., `ruff --select warnings`

I don't think we should overload rule selectors for this. I think I'd prefer a dedicated CLI option like `--severity` to specify the minimum level that is reported.

**Comment by charliermarsh** [comment]
> Thanks for consulting with me :)

Of course! You're driving a lot of the work on the violations API. Definitely want your help figuring these parts out :)

> LSP also supports other levels...

Yeah I've considered this... I wasn't sure if they would really be relevant to us. Like Clippy only has allow, warn, deny, and none. I'm not sold either way.

> I don't think we should overload rule selectors for this. I think I'd prefer a dedicated CLI option like --severity to specify the minimum level that is reported.

Makes sense -- just an example to articulate what I _wasn't_ planning to do on the first pass. I agree that the API you're suggesting is better.

**Comment by not-my-profile** [comment]
> I wasn't sure if they would really be relevant to us.

I am not sure either. I still think

```toml
[tool.ruff.severities]
B = "warning"
```

is better than

```toml
[tool.ruff]
warning = ["B"]
```

because the former is more expressive, e.g.

```toml
[tool.ruff.severities]
B = "warning"
B007 = "error"
```

> Like Clippy only has allow, warn, deny, and none.

These are not defined by clippy ... these are the [Rust lint levels](https://doc.rust-lang.org/rustc/lints/levels.html):

1. allow
2. warn
3. force-warn
4. deny
5. forbid

But what you're saying makes sense ... I think it's ok to start out with just supporting "warning" & "error".

Actually I think we probably should rather model this after clippy since I think it would probably make sense for us to also support "force-warn" and "forbid" (which basically is "force-deny").

So we could have:

```toml
[tool.ruff.levels]
B = "warn"
B007 = "deny"
```

We could actually also deprecate our `ignore` setting in favor of e.g.

```toml
[tool.ruff.levels]
E501 = "allow"
```

The nice thing about the Rust terminology is that "allow" and "deny" are semantic opposites (while "ignore" and "error" aren't).

**Comment by charliermarsh** [comment]
Yeah I generally agree.

My only hesitation around modeling this after Clippy is that the API will be new to most users, whereas `warning` and `error` are more common (e.g. they also map 1:1 to the LSP).

But... it'd be nice to have a unified API for turning errors "on" and "off", _and_ for setting their severities, as with Clippy. (E.g., if we add `[tool.ruff.severities]` but use it in conjunction with `select` and `ignore`, you could turn the B rules "off" _and_ set their severity to "warning", which is strange.)

**Comment by Sawbez** [comment]
@charliermarsh should we add this to a milestone or pin it in some way? Might be useful for someone looking at the repository as an overview. This could also be placed in some sort of `ROADMAP.md`, which we can also update for a version 2 or whatever (I presume version 2 could be adding a formatter, version 3 type hinting? if we are planning to go for those)

**Comment by Agalin** [comment]
I'd really like this to support more than just "error" and "warning".

While discussion in this issue is mostly about editor, ruff is also used in CI and CI apps also use severities. [Gitlab defines 5](https://docs.gitlab.com/ee/ci/testing/code_quality.html#implement-a-custom-tool) and none of them is `warning` nor `error`...

Approach proposed by @not-my-profile would make it possible to handle - if severity list is extended or free form supported.

Why such granularity may be useful in CI?
Because ruff includes rules of really varying importance. If you have an important fix to deploy it may pass review with imports not sorted properly (isort rule) but not with a serious security issue detected by a bandit rule.

**Comment by teucer** [comment]
Having configurable severity levels is important for pipelines. Having gitlab levels (potentially preconfigured) would add a lot of value. 

Right now everything is major. I have added a script to postprocess the generated json file, but it is very hard to decide on the levels.

**Comment by JSpenced** [comment]
Any progress here or when we think this will go into a version?

**Comment by tgross35** [comment]
If user configuration of lint levels is desired, I think it could be reasonable to follow the design specified in Rust's [`manifest-lint` RFC](https://rust-lang.github.io/rfcs/3389-manifest-lint.html), since it gives pretty good reasoning for design choices and was thought through pretty thoroughly. Something like:

```toml
select = ["E", "F", "B"]

[lints]
E123 = "warn"
F456 = { level = "error", priority = 100 }
```

(priority probably isn't really required). Naming the table something like `lints` instead of `severity` has the benefit that it makes sense if you want to add per-lint config that isn't only severity level, e.g.

```toml
[lints.E123]
option-a = "syrup"
ignore = "**/pancakes.py"
```

**Comment by kaddkaka** [comment]
Jenkins CI has a plugin called warnings-ng for reporting issues from various tools. see https://github.com/jenkinsci/warnings-ng-plugin/blob/main/doc/Documentation.md#issues-history-new-fixed-and-outstanding-issues

Unfortunately ruff format is not supported yet. If we could have that it might be able to detect which warnings are new and which are old. This would make it easier to start using ruff in a repo with a lot of warnings by using a baseline.

see https://github.com/jenkinsci/warnings-ng-plugin/blob/main/SUPPORTED-FORMATS.md

**Comment by MusicalNinjaDad** [comment]
> I think having a "info" level in addition to "warning" and "error" could be nice for some lints like PTH (`flake8-use-pathlib`), which i usually enable to remind me and others when we could be using pathlib, but we don't want to strictly enfore that. I'm not sure about other IDEs, but VSCode also supports displaying with an info level.

Personally, I use `--add-noqa` for these kind of info lints. Then I get the info however I look at the codebase, can search for them if e.g. I want to look at migrating to using pathlib, and get in-IDE popups (in VSCode) with the rule details.

Maybe having customisability of --add-noqa would be a solution to this use case? So for example you can specify a rule(set) e.g.:
`ruff --add-noqa PTH` or `ruff --add-noqa D4` etc...?

**Comment by dhruvmanila** [comment]
> Maybe having customisability of --add-noqa would be a solution to this use case? So for example you can specify a rule(set) e.g.:
> `ruff --add-noqa PTH` or `ruff --add-noqa D4` etc...?

This is possible by just combining it with `--select`. So, your examples would become
```
ruff check --select PTH --add-noqa
ruff check --select D4 --add-noqa
```
This should only add `noqa` comments for the selected rules.

**Comment by FichteFoll** [comment]
Since it hasn't been explicitly mentioned yet, I'd also like to point one use case of autofixing. 

For some eslint project, I have turned the most common autofixable lint errors into warnings so that I am aware of them while editing. Since I personally am auto-formatting on save (and fixing all these autofixables at once) I don't need them to be marked as warnings, but not everyone else on the team might be, I still want these to cause the CI lint pass to fail.

Thus, perhaps a way of specifying the level/severity of all autofixable rules at once on a machine-by-machine basis, e.g. by a parameter to the language server, would make sense here.

**Comment by AmeerArsala** [comment]
pls do this. I just want to put a warning on unused imports, nothing else

**Comment by kaddkaka** [comment]
> Jenkins CI has a plugin called warnings-ng for reporting issues from various tools. see https://github.com/jenkinsci/warnings-ng-plugin/blob/main/doc/Documentation.md#issues-history-new-fixed-and-outstanding-issues
> 
> Unfortunately ruff format is not supported yet. If we could have that it might be able to detect which warnings are new and which are old. This would make it easier to start using ruff in a repo with a lot of warnings by using a baseline.
> 
> see https://github.com/jenkinsci/warnings-ng-plugin/blob/main/SUPPORTED-FORMATS.md

Unit support was actually added in #968

**Comment by ShravanSunder** [comment]
It would be great to have severity levels and have this be integrated into ruff vscode plugin

**Comment by Sillocan** [comment]
I would absolutely love warnings in codeclimate rather than errors for some errors such as TODOs

**Comment by beauxq** [comment]
I would like configurable severity levels.

Please don't design it in a way that only works for "error" and "warning". Other levels, like "hint" are also useful.

Other tools that I like that use "hint", don't emit anything at all for "hint" in the CLI output, and only show the "hint" diagnostic in the editor.

Also, I'm worried about how it will be to specify large groups of rules efficiently.
For example, I'd like to be able to say something like:
All of `B` is "warning", except for `B016` which is "error".
And all of `E` is "error", except for `E115` which is "warning".
And the rest of `ALL` is all "hint", except for `PLC0208` which is "warning" and `A001` which is "error".

It would be bad to have to specify hundreds of "hint" lines in `ruff.toml` for something like that.

**Comment by k0pernikus** [comment]
> I would like configurable severity levels.
> 
> Please don't design it in a way that only works for "error" and "warning". Other levels, like "hint" are also useful.

Exactly what [I was looking for](https://github.com/astral-sh/ruff/issues/22801#issue-3842783192).

I am tyring to integrate an `info` vs `error` ruff lints in gitlab's-ci code quality overview.

[Gitlab](https://docs.gitlab.com/ci/testing/code_quality/#code-quality-report-format) supports `info`, `minor`, `major`, `critical`, or `blocker`, yet of course it should not be hardcoded to tools like gitlab, though configuration per `--output-format` would be really great; like mapping internal ruff severities to the tooling's severity if different or find the closest match (or leave for the user to define).

---

I want to generate two different reports:

- `uvx ruff check --output-format=gitlab --output-file=$CQ_REPORT .`
  - this should be major
- `uvx ruff check --isolated --select FIX,TD006 --output-format=gitlab --output-file=$CQ_REPORT . `
  -  this should be info

So please keep in mind what tools are targeted in the report generation, esp. in conjunction with `--isolated`, `--select`, and `--output-format`.

**Comment by k0pernikus** [comment]
Also there should be a different mapping for safe and unsafe fixes:

```shell
 ~  uvx ruff check 
F841 Local variable `foo` is assigned to but never used
  --> portal\gui\tests\login_test.py:18:5
   |
16 | @pytest.mark.unauthenticated
17 | def test_user_can_login(page: Page, user: User):
18 |     foo                =             f"bad code for ruff to complain about         "
   |     ^^^
19 |
20 |     login_page = LoginPage(page, user)
   |
help: Remove assignment to unused variable `foo`

F541 [*] f-string without any placeholders
  --> portal\gui\tests\login_test.py:18:38
   |
16 | @pytest.mark.unauthenticated
17 | def test_user_can_login(page: Page, user: User):
18 |     foo                =             f"bad code for ruff to complain about         "
   |                                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
19 |
20 |     login_page = LoginPage(page, user)
   |
help: Remove extraneous `f` prefix

Found 2 errors.
[*] 1 fixable with the `--fix` option (1 hidden fix can be enabled with the `--unsafe-fixes` option).
```
Both get the severity `major`. If a fix is deemed safe to apply by ruff, it should at most be `minor` or even just `info`.

`$ uvx ruff check --output-format=gitlab`:
```json
[
  {
    "check_name": "F841",
    "description": "F841: Local variable `foo` is assigned to but never used",
    "severity": "major",
    "fingerprint": "84a53cc2ff2feb0",
    "location": {
      "path": "portal\\gui\\tests\\login_test.py",
      "positions": {
        "begin": {
          "line": 18,
          "column": 5
        },
        "end": {
          "line": 18,
          "column": 8
        }
      }
    }
  },
  {
    "check_name": "F541",
    "description": "F541: f-string without any placeholders",
    "severity": "major",
    "fingerprint": "9c5ab6b94e526dcf",
    "location": {
      "path": "portal\\gui\\tests\\login_test.py",
      "positions": {
        "begin": {
          "line": 18,
          "column": 38
        },
        "end": {
          "line": 18,
          "column": 85
        }
      }
    }
  }
]
```

(I want my build pipeline to only fail if human interaction is required. Valid code that is just formatted wrong (and can be fixed automaticaly by a ruff), and rules that have a safe fixer avaiable (that can also be run automatically) should not fail a build.

In the long run I want to run `ruff format` and `run check --fix` automatically within in my pipeline anyway. For now, I want to make sense of the rules before I go full automatic way. Having the the option to map the filter would also help me decide from the code quality section alone if I agree with my rules what should be safe and unsafe fixes.

**Comment by DetachHead** [comment]
personally in my IDE i like to have diagnostics that are likely to cause a runtime crash reported as an error, and less important more style-related diagnostics reported as a warning.however i think it's important for the CLI to have the option to exit with a non-zero code on warnings, because they should always be caught in CI. otherwise your editor will always be full of warnings that weren't introduced by you, causing alarm fatigue.
